controllers/cloud.redhat.com/clowdjobinvocation_controller.go

```
@@ -0,0 +1,374 @@
+			return ctrl.Result{}, err
+		}
+
+		// becuase a CJI can contain > 1 job, we must handle the case
+		// where one job is done and the other is still running
+		fullJobName := fmt.Sprintf("%v-%v-%v", app.Name, job.Name, cji.Name)
+		if contains(cji.Status.Jobs, fullJobName) {
+			continue
+		}
+
+		// We have a match that isn't running and can invoke the job
+		r.Log.Info("Invoking job", "jobinvocation", jobName, "namespace", app.Namespace)
+
+		if err := r.InvokeJob(ctx, &job, &app, &env, &cji); err != nil {
+			r.Log.Info("Job Invocation Failed", "jobinvocation", jobName, "namespace", app.Namespace)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	// Short running jobs may be done by the time the loop is ranged,
+	// so we update again before the reconcile ends
+	r.setCompletedStatus(ctx, &cji)
+	err = r.Client.Status().Update(ctx, &cji)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// InvokeJob is responsible for applying the Job. It also updates and reports
+// the status of that job
+func (r *ClowdJobInvocationReconciler) InvokeJob(ctx context.Context, job *crd.Job, app *crd.ClowdApp, env *crd.ClowdEnvironment, cji *crd.ClowdJobInvocation) error {
+	nn := types.NamespacedName{
+		Name:      fmt.Sprintf("%v-%v-%v", app.Name, job.Name, cji.Name),
+		Namespace: app.Namespace,
+	}
+	j := batchv1.Job{}
+
+	createJobResource(cji, env, nn, job, &j)
+	if err := r.Client.Create(ctx, &j); err != nil {
+		return err
+	}
+	cji.Status.Jobs = append(cji.Status.Jobs, j.ObjectMeta.Name)
+	r.Log.Info("Job Invoked Successfully", "jobinvocation", job.Name, "namespace", app.Namespace)
+
+	return nil
+}
+
+// applyJob build the k8s job resource and applies it from the Job config
+// defined in the ClowdApp
+func createJobResource(cji *crd.ClowdJobInvocation, env *crd.ClowdEnvironment, nn types.NamespacedName, job *crd.Job, j *batchv1.Job) error {
```

It looks like we don't need to return an error here at all